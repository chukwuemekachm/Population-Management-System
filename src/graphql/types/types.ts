import { Request } from 'express-serve-static-core';

import { Prisma } from '../prisma/generated/prisma-client';

// Data Types
export type DateTime = string;
export type ID = string;

// Resolver Types
export type AuthResponseResolver = (parent: AuthResponse) => AuthResponse;
export type UserResolver = (parent: User) => User;
export type LocationResolver = (parent: Location) => Location;

// Common Types
export interface Context {
  request: Request | {};
  prisma: Prisma;
}

export interface AuthResponse {
  token: string;
  user: Partial<User>;
}

// Model Types
export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  name: string;
  maleResidents: number;
  femaleResidents: number;
  totalResidents: number;
  parentLocationId: ID | null;
  subLocations: Location[];
  creator: User;
  createdAt: DateTime;
  updatedAt: DateTime;
}
