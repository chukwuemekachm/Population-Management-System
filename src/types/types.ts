import { Request } from 'express-serve-static-core';

import { Prisma } from '../prisma/generated/prisma-client';

export type DateTime = string;
export type ID = string;

export interface Context {
  request: Request;
  prisma: Prisma;
}

export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  name: string;
  maleResidents: number;
  femaleResidents: number;
  totalResidents: number;
  parentLocationId: ID;
  subLocations: Location[];
  creator: User;
  createdAt: DateTime;
  updatedAt: DateTime;
}
