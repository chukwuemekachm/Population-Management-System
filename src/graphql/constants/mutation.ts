import { gql } from 'apollo-server-express';

export const SIGNUP_MUTATION = gql`
  mutation signup($user: SignupInput) {
    signup(user: $user) {
      user {
        email
        firstName
        lastName
      }
    }
  }
`;

export const LOGIN_MUTATION = gql`
  mutation login($user: LoginInput) {
    login(user: $user) {
      user {
        firstName
        lastName
      }
    }
  }
`;
