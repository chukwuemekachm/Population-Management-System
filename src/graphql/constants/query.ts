import { gql } from 'apollo-server-express';

export const GET_LOCATIONS_QUERY = gql`
  query getLocations($filter: LocationFilterInput) {
    getLocations(filter: $filter) {
      locationName
    }
  }
`;
