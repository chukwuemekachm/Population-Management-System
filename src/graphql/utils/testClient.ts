import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';

import typeDefs from '../schema';
import resolvers from '../resolvers';
import prismaTestMockClient from '../__mocks__/prisma';
import { User } from '../types/types';

export const constructGraphqlTestClient = (user?: Partial<User>) => {
  const testServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: ({ req }) => ({
      request: req,
      prisma: prismaTestMockClient,
      user,
    }),
  });

  return createTestClient(testServer);
};
