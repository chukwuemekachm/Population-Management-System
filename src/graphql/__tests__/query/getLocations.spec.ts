import prismaTestMockClient from '../../__mocks__/prisma';
import { GET_LOCATIONS_QUERY, CREATE_LOCATION_MUTATION } from '../../constants';
import { userInputs } from '../../fixtures/userFixtures';
import { constructGraphqlTestClient } from '../../utils/testClient';
import { User } from '../../types/types';
import { locationInputs } from '../../fixtures/locationFixtures';

let user: Partial<User>;

describe('get locations query', () => {
  beforeAll(async () => {
    const { mutate } = constructGraphqlTestClient();
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
    user = await prismaTestMockClient.createUser(userInputs[1]);
    const payload = {
      variables: {
        location: locationInputs[1],
      },
    };
    await mutate({
      mutation: CREATE_LOCATION_MUTATION,
      ...payload,
    });
  });
  afterAll(async () => {
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
  });

  test('should return locations on the platform', async () => {
    const { query } = constructGraphqlTestClient(user);
    const { errors, data } = await query({
      mutation: GET_LOCATIONS_QUERY,
    });
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  test('should filter locations when filter values are provided', async () => {
    const { query } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        filter: {
          search: 'u',
          offset: 0,
          limit: 3,
          orderBy: 'locationName_ASC',
        },
      },
    };
    const { errors, data } = await query({
      mutation: GET_LOCATIONS_QUERY,
      ...payload,
    });
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  test('should return errors when an invalid filter is supplied', async () => {
    const { query } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        filter: {
          search: 'u',
          offset: 0,
          limit: 3,
          orderBy: 'locationName_BAD',
        },
      },
    };
    const { errors, data } = await query({
      mutation: GET_LOCATIONS_QUERY,
      ...payload,
    });
    expect(errors).toBeDefined();
    expect(data).toBeUndefined();
    expect(errors).toMatchSnapshot();
  });
});
