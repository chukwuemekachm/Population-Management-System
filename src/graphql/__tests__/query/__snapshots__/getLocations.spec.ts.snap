// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get locations query should filter locations when filter values are provided 1`] = `
Object {
  "getLocations": Array [],
}
`;

exports[`get locations query should return errors when an invalid filter is supplied 1`] = `
Array [
  [GraphQLError: Variable "$filter" got invalid value { search: "u", offset: 0, limit: 3, orderBy: "locationName_BAD" }; Expected type LocationOrderBy at value.orderBy; did you mean locationName_ASC or locationName_DESC?],
]
`;

exports[`get locations query should return locations on the platform 1`] = `
Object {
  "getLocations": Array [],
}
`;
